### PR TITLE
nixos/ssh: add variant to not set ForwardX11

### PR DESCRIPTION
**Pull request by SuperSandro2000**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
> A `ForwardX11` in `extraConfig` that cannot easily be removed?

I have it in my user config and it can't be anywhere in the system wide config because then I can't set it based on host.



> should change to use generic `settings` like `services.openssh` does.

Then it should also support setting settings to null to remove them compeltely.

File: nixos/modules/programs/ssh.nix [hunk 1]
```diff
@@ -305,11 +303,8 @@ in
         AddressFamily ${if config.networking.enableIPv6 then "any" else "inet"}
         GlobalKnownHostsFile ${concatStringsSep " " knownHostsFiles}
 
-        ${optionalString cfg.setXAuthLocation ''
-          XAuthLocation ${pkgs.xorg.xauth}/bin/xauth
-        ''}
-
-        ForwardX11 ${if cfg.forwardX11 then "yes" else "no"}
+        ${optionalString cfg.setXAuthLocation "XAuthLocation ${pkgs.xorg.xauth}/bin/xauth"}
+        ${lib.optionalString (cfg.forwardX11 != null) "ForwardX11 ${if cfg.forwardX11 then "yes" else "no"}"}
```
Comment (pbsds) [inline nixos/modules/programs/ssh.nix:307]: I'd prefer not touching XAuthLocation unnecessarily, as reformatting is the bane of `git blame`.

```suggestion
        ${optionalString cfg.setXAuthLocation ''
          XAuthLocation ${pkgs.xorg.xauth}/bin/xauth
        ''}
        ${lib.optionalString (cfg.forwardX11 != null) ''
          ForwardX11 ${if cfg.forwardX11 then "yes" else "no"}
        ''}
```

When templating a config file i believe multiline formatting is better anyway

File: nixos/modules/programs/ssh.nix [hunk 2]
```diff
@@ -52,7 +50,7 @@ in
       };
 
       forwardX11 = mkOption {
-        type = types.bool;
+        type = with lib.types; nullOr bool;
         default = false;
         description = lib.mdDoc ''
           Whether to request X11 forwarding on outgoing connections by default.
```
Comment (pbsds) [inline nixos/modules/programs/ssh.nix:56]: Please document/discuss the `null` behavior on in the description

**Review by pbsds — COMMENTED** [review]
Change looks sane to me, haven't tested though.

> I have it in my user config and it can't be anywhere in the system wide config because then I can't set it based on host.

I do not share this issue. `~/.ssh/config` is loaded before `/etc/ssh/ssh_config` according to `ssh -vvvvvvv`

**Comment by SuperSandro2000** [comment]
> I do not share this issue. `~/.ssh/config` is loaded before `/etc/ssh/ssh_config` according to `ssh -vvvvvvv`

My config is to complex 🤦🏼
I have an include somewhere in my user config which loads the system wide config.

**Comment by pbsds** [comment]
@ofborg test sftpgo sanoid gitolite

>  I have an include somewhere in my user config which loads the system wide config.

I have over 100 hosts and do not do this. I'd love to view your cthulhulean concoction :P

**Pull request merged**